framework/src/media/media_player.c

```diff
@@ -0,0 +1,316 @@
+}
+
+play_result_t media_pause_play(void)
+{
+	PLAY_LOCK();
+
+	if (g_pc.state != PLAY_RUNNING) {
+		PLAY_UNLOCK();
+		return MEDIA_ERROR_INVALID_PARAMETER;
+	}
+
+	char instruction = 0x02;
+	send(g_pc.fd, &instruction, 1, 0);
+
+	g_pc.state = PLAY_PAUSED;
+
+	PLAY_UNLOCK();
+
+	return MEDIA_OK;
+}
+
+play_result_t media_resume_play(void)
+{
+	if (g_pc.state != PLAY_PAUSED) {
+		return MEDIA_ERROR_INVALID_PARAMETER;
+	}
+
+	char instruction = 0x03;
+	send(g_pc.fd, &instruction, 1, 0);
+
+	g_pc.state = PLAY_RUNNING;
+	sem_post(&g_pc.sem_pause);
+
+	return MEDIA_OK;
+}
+
+play_result_t media_set_vol(unsigned char vol)
+{
+	return vol;
+}
+
+unsigned char media_get_vol(void)
+{
+	return 0;
+}
+
+play_result_t media_set_position(int time_ms)
+{
+	return MEDIA_OK;
+}
+
+int media_get_position(void)
+{
+	return 0;
+}
+
+bool media_is_available(void)
+{
+	return true;
+}
+
+bool media_is_playing(void)
+{
+	return (g_pc.state == PLAY_RUNNING);
+}
+
+/**
+ * Internal Functions
+ */
+
+int read_data_pcm_write(char *buffer)
+{
+	int readed = 0;
+	int remain;
+	int ret;
+
+	while (readed < g_pc.buffer_size) {
+		ret = recv(g_pc.fd, buffer + readed, g_pc.buffer_size - readed, 0);		
+		readed += ret;
+		if (ret == 0) {
+			break;
+		}
+	}
+	g_pc.download_size += readed;
+	if (readed) {
+		remain = pcm_bytes_to_frames(g_pc.pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(g_pc.pcmout, buffer + readed - pcm_frames_to_bytes(g_pc.pcmout, remain), remain);
+			if (ret > 0) {
+				remain -= ret;
+			} else {
+				return ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int threadfunc_media_playing(void *args)
+{
+	int ret;
+	char *buffer = NULL;
+	buffer = (char *)malloc(g_pc.buffer_size);
+
+	if (!buffer) {
+		PLAY_LOCK();
+		g_pc.state = PLAY_AVAILABLE;
+		PLAY_UNLOCK();
+		return PLAY_ERROR_MALLOC;
+	}
+
+	while (g_pc.thread_running) {
+		PLAY_LOCK();
+		switch (g_pc.state) {
+		case PLAY_RUNNING:
+			ret = read_data_pcm_write(buffer);
+			if (ret == 0) {
+				g_pc.state = PLAY_DRAINING;
+			} else if (ret < 0) {
+				g_pc.state = PLAY_STOPPING;
+			}
+			break;
+
+		case PLAY_PAUSED:
+			PLAY_UNLOCK();
+			sem_wait(&g_pc.sem_pause);
+			PLAY_LOCK();
+			break;
+
+		case PLAY_STOPPING:
+			g_pc.thread_running = false;
+			break;
+
+		case PLAY_DRAINING:
+			pcm_drain(g_pc.pcmout);
+			g_pc.state = PLAY_STOPPING;
+
+		default:
+			g_pc.thread_running = false;
+			g_pc.state = PLAY_STOPPING;
+			break;
+		}
+		PLAY_UNLOCK();
+
+		// For context-switch
+		sched_yield();
+	}
+
+	free(buffer);
+
+	PLAY_LOCK();
+	g_pc.state = PLAY_AVAILABLE;
+	PLAY_UNLOCK();
+
+	return 0;
+}
+
+play_result_t media_play_init(void)
+{
+	if (g_pc.state == PLAY_ERROR) {
+		memset(&g_pc, 0, sizeof(g_pc));
+
+		g_pc.state = PLAY_AVAILABLE;
+
+		pthread_mutexattr_init(&g_pc.mutex_attr_play);
+		pthread_mutexattr_settype(&g_pc.mutex_attr_play, PTHREAD_MUTEX_NORMAL);
+		pthread_mutex_init(&g_pc.mutex_play, &g_pc.mutex_attr_play);
+
+		g_pc.fd = -1;
+
+		struct pcm_config config;

```

Comment:
Here is wrong.
Assume that we gonna play 2 file one is 16K sample rate and last one is 44K.
User will call function as below
media_play_init();
if (media_prepare()) {
     play(FIRST_FILE_FD);
}
stop();
if (media_prepare()) {
    play(SECOND_FILE_FD);
}
stop();
Then how we can set configuration value for 2nd file?
Also buffer size also should be changed.
So configuration should be moved to prepare level.
And at that time it should check header if it is wav file.

Reply:
I think we should think about the place where we close the pcm device, the device is not closed anywhere and it is a leak for the device because no application using tinyalsa apis directly can call open pcm device after mediaplayer is initialised since pcm_open will return errors.

Reply:
As we discussed, One option is to move pcm_open at the media_play function and add pcm_close at the bottom of thread function. Is there another way?